### DISCUSSION TITLE
Can't debug packaging project for unpackaged WinUI 3 app (DEP0800)

### DISCUSSION THREAD

**Discussion by Fritierkelle** [microsoft/microsoft-ui-xaml#9386] [Q&A]
Hi,
I added a packaging project to my WinUI3 unpackaged app and set it as startup project (VS 2022/ Windows 11).
I can't debug and get this error (same error when I try to install the MsixBundle):

> Creating a new clean layout...
Copying files: Total 12 mb to layout...
Checking whether required frameworks are installed...
Framework: Microsoft.WindowsAppRuntime.1.2/X64, app package version 5001.58.448.0 is not currently installed.
Installing missing frameworks...
Registering the application to run from layout...

> DEP0700: Registration of the app failed. [0x80073CF3] Windows cannot install package f7acecac-ea75-4558-b8d3-4d262895a266_1.0.5.0_x64__a1j3z9m6cv6fy 
because this package depends on a framework that could not be found. Provide the framework "Microsoft.WindowsAppRuntime.1.2" published by 
"CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US", with neutral or x64 processor architecture and minimum version 
5001.58.448.0, along with this package to install.
========== Deploy: 1 succeeded, 0 failed, 0 skipped ==========
========== Deploy completed at 3:10 AM and took 03.355 seconds ==========

I also had another error:

> DEP0800: The required framework "C:\Users\xxx\source\repos\cpp\UI_Test 3\packages\Microsoft.WindowsAppSDK.1.2.230118.102\build..\tools\MSIX\win10-x64\Microsoft.WindowsAppRuntime.1.2.msix" failed to install. [0x80073D06] Windows cannot install package Microsoft.WindowsAppRuntime.1.2_2000.747.1945.0_x64__8wekyb3d8bbwe because it has version 2000.747.1945.0. A higher version 2000.802.31.0 of this package is already installed.

I had all packages mentioned in  the errors installed and uninstalled but the errors run in circle.

These are my appxpackages:

> (get-appxpackage micro*win*appruntime*).packagefullname
Microsoft.WindowsAppRuntime.1.5_5001.58.448.0_x64__8wekyb3d8bbwe
Microsoft.WindowsAppRuntime.1.5_5001.58.448.0_x86__8wekyb3d8bbwe
MicrosoftCorporationII.WinAppRuntime.Main.1.5_5001.58.448.0_x64__8wekyb3d8bbwe
Microsoft.WinAppRuntime.DDLM.5001.58.448.0-x6_5001.58.448.0_x64__8wekyb3d8bbwe
Microsoft.WinAppRuntime.DDLM.5001.58.448.0-x8_5001.58.448.0_x86__8wekyb3d8bbwe
MicrosoftCorporationII.WinAppRuntime.Singleton_5001.58.448.0_x64__8wekyb3d8bbwe
Microsoft.WindowsAppRuntime.1.1_1005.616.1651.0_x64__8wekyb3d8bbwe
Microsoft.WindowsAppRuntime.1.1_1005.616.1651.0_x86__8wekyb3d8bbwe
Microsoft.WindowsAppRuntime.1.3_3000.934.1904.0_x64__8wekyb3d8bbwe
Microsoft.WindowsAppRuntime.1.3_3000.934.1904.0_x86__8wekyb3d8bbwe
Microsoft.WindowsAppRuntime.1.4_4000.1136.2333.0_x86__8wekyb3d8bbwe
Microsoft.WindowsAppRuntime.1.4_4000.1136.2333.0_x64__8wekyb3d8bbwe
Microsoft.WindowsAppRuntime.1.2_2000.802.31.0_x86__8wekyb3d8bbwe

- WinUI 3 Nuget updated to 1.5
- 1.1 (Microsoft Family currently depends on the framework...)
- 1.3 (Microsoft.ScreenSketch currently...)
- 1.4 (Microsoft.YourPhone currently...)

**Is it a bug?**
(The app runs flawlessly without a packaging project.)

**Comment by DarranRowe** [comment]
Have you updated the NuGet packages for the packaging project?
When you have a packaging project, it obtains the package dependencies from the NuGet packaged referenced to the packaging project. So, as an example:

![Screenshot 2024-03-03 114142](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/c3eecbfc-d66d-45fa-b413-0a79b644f626)

You can see from this, the packaging project references 1.5. You really want to keep this in line with the application itself otherwise problems will occur.

If you read the error message that you posted, it is telling you that the packaging project is trying to install 1.2 but there is a different version of 1.2 installed. With the way the Windows App SDK is versioned, the packaging project must be set to 1.5 if the application uses the Windows App SDK 1.5.

**Reply by Fritierkelle** [reply]
Yes,
That's what I've tried to solve primarily.

(BuildTools & WindowsAppSDK are both up-to-date and installed on the app and packaging project.)
![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/62b0177f-3496-48ef-a575-8e929f20045b)
![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/5b939b63-ceb2-4ed6-8116-e5c63999a1c5)

And this has nothing to do with [deployment configurations](https://learn.microsoft.com/en-us/windows/apps/package-and-deploy/deploy-overview#more-info-about-self-contained-deployment)? 

I thought I could make it [self-contained](https://learn.microsoft.com/en-us/windows/apps/package-and-deploy/self-contained-deploy/deploy-self-contained-apps) and it could fix it?!

**Reply by DarranRowe** [reply]
You shouldn't need to make it self contained. It would copy the Windows App SDK runtime into the package and use Registration Free WinRT, this would stop you from using features like App Notifications and you would have to update your package each time the Windows App SDK is updated.
This also doesn't have anything to do with deployment configuration. When you are working with the packaging project, Visual Studio updates the manifest with the correct version of the Windows App SDK.

So, let's follow that path.
First, have you modified the Package.appxmanifest file at all? In particular, the Dependencies element should look similar to the following.

```
  <Dependencies>
    <TargetDeviceFamily Name="Windows.Universal" MinVersion="10.0.19041.0" MaxVersionTested="10.0.19041.0" />
    <TargetDeviceFamily Name="Windows.Desktop" MinVersion="10.0.19041.0" MaxVersionTested="10.0.19041.0" />
  </Dependencies>
```
You could have just Windows.Desktop, the MinVersion and MaxVersionTested could be different, but the important thing is that it should only contain TargetDeviceFamily entries. If you do add any PackageDependency entries, they must be entries that are not automatically added by the project.

Secondly, after you build the project, find the generated AppxManifest.xml file. There should be a platform/configuration directory under the packaging projects bin directory that contains the file. If you don't exactly know where that directory is, you can always right click on the packaging project directory and select Open Folder in File Explorer.

![Screenshot 2024-03-03 191724](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/99c38fba-14f0-4ebc-a2fc-63c005cbf31e)

![Screenshot 2024-03-03 191438](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/f461d30d-6782-441b-bf2f-6df18ff232d2)

The example above was built with the x64/Debug configuration. This file should contain the generated dependencies in the Dependencies element.

```
  <Dependencies>
    <TargetDeviceFamily Name="Windows.Universal" MinVersion="10.0.19041.0" MaxVersionTested="10.0.19041.0" />
    <TargetDeviceFamily Name="Windows.Desktop" MinVersion="10.0.19041.0" MaxVersionTested="10.0.19041.0" />
    <PackageDependency Name="Microsoft.WindowsAppRuntime.1.5" MinVersion="5001.58.448.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
    <PackageDependency Name="Microsoft.VCLibs.140.00.Debug" MinVersion="14.0.30704.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
    <PackageDependency Name="Microsoft.VCLibs.140.00.Debug.UWPDesktop" MinVersion="14.0.30704.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
  </Dependencies>
```

It takes the entry from the Package.appxmanifest file and adds any extra information from the project itself. In this case, it obtains this information from the Windows App SDK NuGet package reference. So does this look correct? If it doesn't, try completely cleaning the project and rebuilding it.

If you do want to try making the project self contained, make sure to add WindowsAppSDKSelfContained to both your project file (.vcxproj or .csproj) as well as the packaging project (.wapproj) file.

**Reply by Fritierkelle** [reply]
I really only added the packaging project, made it the StartUp project and that's it. (The exact same steps like in my other Win32 app.)

I already deleted the bin folder yesterday but it didn't change anything.

AppxManifest:
![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/f980346d-9d58-426b-a152-25c1cac40ad8)

Package.appxManifest:
![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/06cb9848-e6b1-4e7c-a030-5550f86053b6)

Configuration Manager:
![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/5c052d5b-bb7c-4628-8761-d540e85a8045)

Output:
![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/40966643-73e0-4de4-8389-fcf7714b8954)

![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/d5420964-6698-4c1e-a51c-bff1fc20d22c)

![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/4f20accf-6f88-4dd4-a7d4-64e30827b1be)

I would probably try framework-dependent first?!

**Reply by DarranRowe** [reply]
Did you notice, in the generated AppxManifest.xml file, you have two versions of the Windows App Runtime? One of them is WindowsAppRuntime.1.2, and this is what the error is centring around. What's more, it is an impossible version, 1.2 should always have a 2000 version number.
That is definitely where the problem is, and the big question then is where is it getting this from. Could you check the individual project NuGet references? You can do this by right clicking on the project and selecting Manage NuGet Packages. I just want to be sure that the references are consistent across all projects.

![Screenshot 2024-03-04 011742](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/76b574b4-d0e7-4e89-8a90-ecf102f887f2)

**Reply by Fritierkelle** [reply]
I can't explain what went wrong. I just cleaned up the packages via shell but after the error message(s), that's all.
(I started the project a long time ago. Started with VS2019 and it's 2022 now, a lot of updates later! Can it be faulty due to the registry, under DEP0800 people report this?!)

![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/7b8d8e87-2797-49eb-94d0-c0a0da5607dc)

**Reply by DarranRowe** [reply]
Well, do you have the same problem if you create a new WinUI 3 application?

There are a couple of possibilities, especially if the project is older. I do remember there being some bugs in Visual Studio itself when handling NuGet package installation.
Information regarding the import of NuGet packages are written to the project file (.vcxproj or .csproj). C++ projects in particular can be troublesome because they don't reference the NuGet package directly. The package is copied from the NuGet cache to the C++ solution directory, this directory is then reference in the C++ project file.

For illustration purposes, this is the package directory in a C++ project's solution directory.

![Screenshot 2024-03-04 141031](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/2d128429-7692-4ecf-9dc8-a302baedf024)

The following is the references placed in the .vcxproj file.

The top of the file has imports similar to:
```
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="15.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Import Project="..\..\packages\Microsoft.WindowsAppSDK.1.5.240227000\build\native\Microsoft.WindowsAppSDK.props" Condition="Exists('..\..\packages\Microsoft.WindowsAppSDK.1.5.240227000\build\native\Microsoft.WindowsAppSDK.props')" />
  <Import Project="..\..\packages\Microsoft.Windows.SDK.BuildTools.10.0.22621.3233\build\Microsoft.Windows.SDK.BuildTools.props" Condition="Exists('..\..\packages\Microsoft.Windows.SDK.BuildTools.10.0.22621.3233\build\Microsoft.Windows.SDK.BuildTools.props')" />
  <Import Project="..\..\packages\Microsoft.Windows.CppWinRT.2.0.240111.5\build\native\Microsoft.Windows.CppWinRT.props" Condition="Exists('..\..\packages\Microsoft.Windows.CppWinRT.2.0.240111.5\build\native\Microsoft.Windows.CppWinRT.props')" />
```

The end of the file has imports similar to:

```
  <ImportGroup Label="ExtensionTargets">
    <Import Project="..\..\packages\Microsoft.Windows.ImplementationLibrary.1.0.240122.1\build\native\Microsoft.Windows.ImplementationLibrary.targets" Condition="Exists('..\..\packages\Microsoft.Windows.ImplementationLibrary.1.0.240122.1\build\native\Microsoft.Windows.ImplementationLibrary.targets')" />
    <Import Project="..\..\packages\Microsoft.Windows.CppWinRT.2.0.240111.5\build\native\Microsoft.Windows.CppWinRT.targets" Condition="Exists('..\..\packages\Microsoft.Windows.CppWinRT.2.0.240111.5\build\native\Microsoft.Windows.CppWinRT.targets')" />
    <Import Project="..\..\packages\Microsoft.Windows.SDK.BuildTools.10.0.22621.3233\build\Microsoft.Windows.SDK.BuildTools.targets" Condition="Exists('..\..\packages\Microsoft.Windows.SDK.BuildTools.10.0.22621.3233\build\Microsoft.Windows.SDK.BuildTools.targets')" />
    <Import Project="..\..\packages\Microsoft.WindowsAppSDK.1.5.240227000\build\native\Microsoft.WindowsAppSDK.targets" Condition="Exists('..\..\packages\Microsoft.WindowsAppSDK.1.5.240227000\build\native\Microsoft.WindowsAppSDK.targets')" />
  </ImportGroup>
  <Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">
    <PropertyGroup>
      <ErrorText>This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
    </PropertyGroup>
    <Error Condition="!Exists('..\..\packages\Microsoft.Windows.ImplementationLibrary.1.0.240122.1\build\native\Microsoft.Windows.ImplementationLibrary.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Microsoft.Windows.ImplementationLibrary.1.0.240122.1\build\native\Microsoft.Windows.ImplementationLibrary.targets'))" />
    <Error Condition="!Exists('..\..\packages\Microsoft.Windows.CppWinRT.2.0.240111.5\build\native\Microsoft.Windows.CppWinRT.props')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Microsoft.Windows.CppWinRT.2.0.240111.5\build\native\Microsoft.Windows.CppWinRT.props'))" />
    <Error Condition="!Exists('..\..\packages\Microsoft.Windows.CppWinRT.2.0.240111.5\build\native\Microsoft.Windows.CppWinRT.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Microsoft.Windows.CppWinRT.2.0.240111.5\build\native\Microsoft.Windows.CppWinRT.targets'))" />
    <Error Condition="!Exists('..\..\packages\Microsoft.Windows.SDK.BuildTools.10.0.22621.3233\build\Microsoft.Windows.SDK.BuildTools.props')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Microsoft.Windows.SDK.BuildTools.10.0.22621.3233\build\Microsoft.Windows.SDK.BuildTools.props'))" />
    <Error Condition="!Exists('..\..\packages\Microsoft.Windows.SDK.BuildTools.10.0.22621.3233\build\Microsoft.Windows.SDK.BuildTools.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Microsoft.Windows.SDK.BuildTools.10.0.22621.3233\build\Microsoft.Windows.SDK.BuildTools.targets'))" />
    <Error Condition="!Exists('..\..\packages\Microsoft.WindowsAppSDK.1.5.240227000\build\native\Microsoft.WindowsAppSDK.props')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Microsoft.WindowsAppSDK.1.5.240227000\build\native\Microsoft.WindowsAppSDK.props'))" />
    <Error Condition="!Exists('..\..\packages\Microsoft.WindowsAppSDK.1.5.240227000\build\native\Microsoft.WindowsAppSDK.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Microsoft.WindowsAppSDK.1.5.240227000\build\native\Microsoft.WindowsAppSDK.targets'))" />
  </Target>
</Project>
```

This is problematic because C++ projects have the package information independent of the references. If Visual Studio crashes during the removal of a NuGet package then it is possible that the package could be removed from packages.config but the references may not be removed from the .vcxproj file itself. I have certainly seen a project end up with doubled up references in the .vcxproj file due to this.
Testing this out in a newly created project should help confirm this or rule it out.

**Reply by Fritierkelle** [reply]
Ok,
I will test it tomorrow.
I took an old unpackaged test project and the WAP exited immediately with the following message:

The target process exited without raising a CoreCLR started event. Ensure that the target process is configured to use .NET Core. This may be expected if the target process did not run on .NET Core.
The program '[9920] WinUI3_Unpackaged_Cpp_Blank.exe' has exited with code 2147942450 (0x80070032).

(I'm not shure, if I removed to much.)

**Reply by Fritierkelle** [reply]
- I cleaned the components by Windows 11 > settings > apps.
- Reinstalled the runtimes and VS asked for 1.2_2000.747.1945.0.
- This time I installed it by the Redistributable .zip file, version number is in the MSIX.inventory file (installed 3 of 4 files [not only the WindowsAppRuntimeInstall.exe] except the singleton because it said that a higher version was already installed.)

![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/9c79960f-3b85-4f45-b2f5-d063b56198de)

I tested in my old app, created also a new app with a WAP project.
The runtime errors are gone but I get the new message:

> The target process exited without raising a CoreCLR started event. Ensure that the target process is configured to use .NET Core. This may be expected if the target process did not run on .NET Core.
The program '[16940] CPP_Unpackaged_WAP_TestProject.exe' has exited with code 2147942450 (0x80070032).

I already checked that I have all needed components installed. Any idea?
(The documentation says nothing about .NET core for WAP or winrt/c++)

Do I have to change something here?
![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/01879047-e41b-4b27-a7d4-0dd494df5ba8)

**Reply by DarranRowe** [reply]
Yes, if your application is a C++ application then you really want both of the entries under Debugger type to be Native Only.

![Screenshot 2024-03-05 184006](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/a1332cf1-2f71-4012-90ab-a4f4fe416511)

That is, unless you are hosting the .NET.

**Reply by Fritierkelle** [reply]
I am somehow a bit lost.

![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/39a5bd45-59f6-432c-b97a-6d29bb82e2ec)

I have tried all combinations. You really can't find any information about it on the net.
I followed the instructions and double checked the components.
https://learn.microsoft.com/en-us/windows/msix/desktop/desktop-to-uwp-packaging-dot-net

**Sorry, 
But is it even possible to package an unpackaged WinUI3 c++ app, has it been done successfully?**

**Reply by DarranRowe** [reply]
Yes it has. Given the debugger output in your last screenshot, I would guess you have WindowsPackageType set to None in the .vcxproj file. This causes the project to compile in a source file that automatically call [MddBootstrapInitialize2](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/win32/mddbootstrap/nf-mddbootstrap-mddbootstrapinitialize2). You can see what the call itself defaults to by inspecting the MddBootstrapAutoInitializer.cpp file which is part of the Windows App SDK NuGet package. But it defaults to specifying the current version as the version of the SDK to bootstrap with no options.
Since MddBootstrapInitialize2 is only used by unpackaged applications to reference the Windows App Runtime, by default it just errors out. There are two ways of dealing with this. First is just not using WindowsPackageType in the .vcxproj file. This means that you would have to directly call MddBootstrapInitialize2 when needed. But there is a better way. The second option is using the fact that MddBootstrapInitialize2 has some [options](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/win32/mddbootstrap/ne-mddbootstrap-mddbootstrapinitializeoptions#constants). One of them is to tell MddBootstrapInitialize2 to do nothing if the application is packaged. The way to enable these options, and in general how the bootstrapper works, is [documented](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/use-windows-app-sdk-run-time#behind-the-scenes-and-opting-out-of-automatic-module-initialization).

```
  <PropertyGroup Label="Globals">
    <CppWinRTOptimized>true</CppWinRTOptimized>
    <CppWinRTRootNamespaceAutoMerge>true</CppWinRTRootNamespaceAutoMerge>
    <MinimalCoreWin>true</MinimalCoreWin>
    <ProjectGuid>{772fb297-9de2-42e1-80c3-021526001dac}</ProjectGuid>
    <ProjectName>TestWinUI3</ProjectName>
    <RootNamespace>TestWinUI3</RootNamespace>
    <!--
      $(TargetName) should be same as $(RootNamespace) so that the produced binaries (.exe/.pri/etc.)
      have a name that matches the .winmd
    -->
    <TargetName>$(RootNamespace)</TargetName>
    <DefaultLanguage>en-US</DefaultLanguage>
    <MinimumVisualStudioVersion>16.0</MinimumVisualStudioVersion>
    <AppContainerApplication>false</AppContainerApplication>
    <AppxPackage>false</AppxPackage>
    <ApplicationType>Windows Store</ApplicationType>
    <ApplicationTypeRevision>10.0</ApplicationTypeRevision>
    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
    <WindowsTargetPlatformMinVersion>10.0.17763.0</WindowsTargetPlatformMinVersion>
    <UseWinUI>true</UseWinUI>
    <!-- Used to enable automatic bootstrapping of the Windows App SDK, will fail if the application is packaged. -->
    <WindowsPackageType>None</WindowsPackageType>
    <!-- Does the equivalend of calling MddBootstrapInitialize2 with MddBootstrapInitializeOptions_OnPackageIdentity_NOOP. -->
    <!-- This allows this executable to serve as the binary for both a packaged and unpackaged application, without directly callind MddBootstrapInitialize2. -->
    <WindowsAppSDKBootstrapAutoInitializeOptions_OnPackageIdentity_NoOp>true</WindowsAppSDKBootstrapAutoInitializeOptions_OnPackageIdentity_NoOp>
  </PropertyGroup>
```

The above is a snippet of a test .vcxproj file with this option set.

![Screenshot 2024-03-05 225949](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/c77f0bec-c1e0-40e9-bbb3-7205c4197a65)

You should see in Solution Explorer that the package is set as the startup project. The exact same project works without changes if you just set the project itself as the startup project.

![Screenshot 2024-03-05 230007](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/bc76385d-dcb5-4354-aa1b-48ed94ef4031)

It does work, but it is a bit fiddly getting some of the pieces put together. Here is a small example with only minor edits.

[TestWinUI3.zip](https://github.com/microsoft/microsoft-ui-xaml/files/14502518/TestWinUI3.zip)

**Reply by Fritierkelle** [reply]
Oh ok,
Thank you. 
(Yes, that makes sense.)

PS: It all seems extremely time-consuming somehow. How about simply using a switch (packaged/unpackaged/ framework-dependent/ self-contained)?  :)

**Comment by minfrin** [comment]
I think I am suffering the same problem, but not sure (much unix experience, not much windows).

```
PS C:\Users\minfrin\source\redwax-signtext\src\windows\Redwax SignText Packaging\AppPackages> Add-AppPackage -Path ".\Redwax SignText_1.0.0.0_Win32_Debug_Test\Redwax SignText_1.0.0.0_Win32_Debug.msix" -AllowUnsigned
Add-AppPackage : Deployment failed with HRESULT: 0x80073CF3, Package failed updates, dependency or conflict validation.
Windows cannot install package c78bf4e4-e4e5-44f1-97dd-69e0f4a169db_1.0.0.0_x86__enwe9x4v0qrtw because this package
depends on a framework that could not be found. Provide the framework "Microsoft.WindowsAppRuntime.1.3" published by
"CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US", with neutral or x86 processor
architecture and minimum version 3000.882.2207.0, along with this package to install.
NOTE: For additional information, look for [ActivityId] 2efe139a-721b-0001-2963-0b2f1b72da01 in the Event Log or use
the command line Get-AppPackageLog -ActivityID 2efe139a-721b-0001-2963-0b2f1b72da01
At line:1 char:1
+ Add-AppPackage -Path ".\Redwax SignText_1.0.0.0_Win32_Debug_Test\Redw ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (C:\Users\minfri...in32_Debug.msix:String) [Add-AppxPackage], IOException
    + FullyQualifiedErrorId : DeploymentError,Microsoft.Windows.Appx.PackageManager.Commands.AddAppxPackageCommand
```

When trying to install the MSIX for testing, I get the same failure: "package depends on a framework that could not be found". How did the MSIX build successfully without error when the dependency was missing?

I understand that the MSIX is self contained, but maybe it's not? What stops end users being sent into dependency hell just installing a package? What I thought I understood and the behaviour before me don't match and I am lost.

In my case, the properties look like this:

```
  <PropertyGroup Label="Globals">
    <CppWinRTOptimized>true</CppWinRTOptimized>
    <CppWinRTRootNamespaceAutoMerge>true</CppWinRTRootNamespaceAutoMerge>
    <MinimalCoreWin>true</MinimalCoreWin>
    <ProjectGuid>{2779dda1-3214-4c35-84d2-454f2e542cc0}</ProjectGuid>
    <ProjectName>Redwax SignText</ProjectName>
    <RootNamespace>Redwax_SignText</RootNamespace>
    <!--
      $(TargetName) should be same as $(RootNamespace) so that the produced binaries (.exe/.pri/etc.)
      have a name that matches the .winmd
    -->
    <TargetName>$(RootNamespace)</TargetName>
    <DefaultLanguage>en-US</DefaultLanguage>
    <MinimumVisualStudioVersion>16.0</MinimumVisualStudioVersion>
    <AppContainerApplication>false</AppContainerApplication>
    <AppxPackage>true</AppxPackage>
    <ApplicationType>Windows Store</ApplicationType>
    <ApplicationTypeRevision>10.0</ApplicationTypeRevision>
    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
    <WindowsTargetPlatformMinVersion>10.0.17763.0</WindowsTargetPlatformMinVersion>
    <UseWinUI>true</UseWinUI>
    <EnableMsixTooling>true</EnableMsixTooling>
  </PropertyGroup>
```

**Reply by DarranRowe** [reply]
This is more of the case that when you install the package through the command line like this, Windows will not automatically grab the dependencies from the store.
You need to provide the dependencies alongside the package by using the -DependencyPath command line option of [Add-AppxPackage](https://learn.microsoft.com/en-us/powershell/module/appx/add-appxpackage?view=windowsserver2022-ps). The Windows App SDK packages can be found in the NuGet package itself. For a C++ project, the extracted NuGet package can be found in the packages directory which should be in the same directory as your .sln file.

![Screenshot 2024-03-09 213011](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/71ab813b-490a-4701-be69-98ec19b59b17)

The packages can be found under the Windows App SDK directory in the tools\MSIX\win10-$(Platform) directory.

![Screenshot 2024-03-09 213128](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/3e5ea17e-6946-451f-aec8-fc915ce8fef8)

![Screenshot 2024-03-09 213258](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/c2acbb97-d25f-4446-bca1-dd285f38b49a)

The package you need to provide would be Microsoft.WindowsAppRuntime.1.3.msix. I showed an example using 1.5, but the directory layout is the same. You can install DDLM if you wish, and this will match your application's platform. But if you want to install the singleton and main packages, try to match the system's platform.

If you need to provide the VC Redistributable, you can find that under Program Files (x86) on your development machine. Remember, since you are trying to install a debug package, the VC Runtime isn't redistributable. This will only install alongside Visual C++ or the Windows SDK. The release version is redistributable and you should be able to find versions of it available to download from Microsoft.

![Screenshot 2024-03-09 213940](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/e26316ff-26c4-419e-83b5-68edd29d2caf)

To deploy this via command line, you would then really do the following.

1) Verify what you need. The generated AppxManifest.xml will contain all of this information.

```
  <Dependencies>
    <TargetDeviceFamily Name="Windows.Universal" MinVersion="10.0.19041.0" MaxVersionTested="10.0.19041.0" />
    <TargetDeviceFamily Name="Windows.Desktop" MinVersion="10.0.19041.0" MaxVersionTested="10.0.19041.0" />
    <PackageDependency Name="Microsoft.WindowsAppRuntime.1.5" MinVersion="5001.58.448.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
    <PackageDependency Name="Microsoft.VCLibs.140.00.Debug.UWPDesktop" MinVersion="14.0.30704.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
    <PackageDependency Name="Microsoft.VCLibs.140.00.Debug" MinVersion="14.0.30704.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
  </Dependencies>
```

This is from a test project. From this, you can see that it declares a dependency on Microsoft.WindowsAppRuntime.1.5, Microsoft.VCLibs.140.00.Debug.UWPDesktop and Microsoft.VCLibs.140.00.Debug. I will use the x86/debug version to match your situation.

2) Figure out where all of the dependencies are. I'm going to gather everything into a single directory for ease of deployment, but you don't have to.

![Screenshot 2024-03-09 215056](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/c56dc9d6-ea17-4236-a012-ba74ee26a229)

![Screenshot 2024-03-09 215226](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/7cf47312-bc6d-47b7-985b-e4962bf0be33)

![Screenshot 2024-03-09 215318](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/b84c7a1f-0b8c-4949-a471-8ff3e0e1394e)

These, along with the application package are then copied alongside the application package.

![Screenshot 2024-03-09 215504](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/e15469e9-d7a3-4869-ae83-c8d656e2fa24)

The directory where everything is stored doesn't matter here. I just already had this directory set up for network installs.

3) Install the package along with the dependencies.

![Screenshot 2024-03-09 215905](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/39a637f6-261f-4fc3-a638-fa4bd76cc1f3)

If this doesn't work for you then please comment back.

**Reply by Fritierkelle** [reply]
So I gave up on it for now and wanted to tackle it again in a month, it's really extremely time-consuming.

Don't know if this helps, but I've had problems all along because there was always a dependency on 1.2, even though I usually upgrade to a stable version.

I'm really not the senior specialist. When I cleaned the packages and components, 1.2 kept coming back.
I found a post on [StackOverFlow](https://stackoverflow.com/questions/22909506/this-project-references-nuget-packages-that-are-missing-on-this-computer) and removed the following line from my vcxproj, now the old version is finally gone from the packages folder. 

`<Error Condition="!Exists('..\packages\Microsoft.WindowsAppSDK.1.2.230118.102\build\native\Microsoft.WindowsAppSDK.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\Microsoft.WindowsAppSDK.1.2.230118.102\build\native\Microsoft.WindowsAppSDK.targets'))" />`

![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19283500/18e735cd-1443-4f99-81dc-30f0ba4ea76e)

Regards
